ddsrecorder_participants/src/cpp/recorder/mcap/McapHandler.cpp

```diff
@@ -948,71 +1010,98 @@ void McapHandler::rewrite_schemas_nts_()
+    if (type_identifier)
+    {
+        type_object =
+                eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_name, true);
+    }
+
+    // If complete not found, try with minimal
+    if (!type_object)
+    {
+        type_identifier = eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_identifier(
+            type_name, false);
         if (type_identifier)
         {
-            type_object =
-                    eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_name, true);
+            type_object = eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_name,
+                            false);
         }
+    }
 
-        // If complete not found, try with minimal
-        if (!type_object)
-        {
-            type_identifier = eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_identifier(
-                type_name, false);
-            if (type_identifier)
-            {
-                type_object = eprosima::fastrtps::types::TypeObjectFactory::get_instance()->get_type_object(type_name,
-                                false);
-            }
-        }
+    // Store dynamic type in dynamic_types_
+    store_dynamic_type_(type_identifier, type_object, type_name);
+}
+
+void McapHandler::store_dynamic_type_(
+        const eprosima::fastrtps::types::TypeIdentifier* type_identifier,
+        const eprosima::fastrtps::types::TypeObject* type_object,
+        const std::string& type_name)
+{
+    if (type_identifier != nullptr && type_object != nullptr)
+    {
+        DynamicType dynamic_type;
+        dynamic_type.type_name(type_name);
+        dynamic_type.type_information(utils::base64_encode(serialize_type_identifier_(type_identifier)));
+        dynamic_type.type_object(utils::base64_encode(serialize_type_object_(type_object)));
 
-        // Serialize dynamic type in a string and store in map
-        store_dynamic_type_(type_identifier, type_object, type_name, dynamic_types);
+        dynamic_types_.dynamic_types().push_back(dynamic_type);
     }
+}
+
+void McapHandler::serialize_dynamic_types_()
+{
+    // Serialize dynamic types collection using CDR
+    eprosima::fastdds::dds::TypeSupport type_support(new DynamicTypesCollectionPubSubType());
+    eprosima::fastrtps::rtps::SerializedPayload_t serialized_payload =
+            eprosima::fastrtps::rtps::SerializedPayload_t(
+        type_support.get_serialized_size_provider(&dynamic_types_)());
+    type_support.serialize(&dynamic_types_, &serialized_payload);
+
+    // Recalculate storage_dynamic_types_ when serializing dynamic_types_
+    storage_dynamic_types_ = MCAP_ATTACHMENT_OVERHEAD + serialized_payload.length;
+}
 
+void McapHandler::write_attachment_()
+{
     // Serialize dynamic types collection using CDR
     eprosima::fastdds::dds::TypeSupport type_support(new DynamicTypesCollectionPubSubType());
```

Comment:
I believe this code is the same as `serialize_dynamic_types`. Please reuse the function.

Reply:
I haven't stored the `serialized_payload` as a class member, so it's not accessible for writing the attachment if I simply perform serialize_dynamic_types_() within this function. Conversely, keeping `serialized_payload` as a member was causing numerous issues related to memory deallocation